docs/storage/gcsupload/gcsupload.go

```diff
@@ -0,0 +1,162 @@
+//  URL: https://storage.googleapis.com/gcs-cli-test/birthdayPic.jpg
+//  Size: 865096
+//  MD5: 5b6c7b4aed837e8ed0f9950564a10b32
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	"cloud.google.com/go/storage"
+
+	"golang.org/x/net/context"
+)
+
+func envOrFatal(envKey, fallback string) string {
+	retr := os.Getenv(envKey)
+	if retr == "" {
+		if fallback != "" {
+			return fallback
+		}
+		log.Fatalf("%q should be set", envKey)
+	}
+	return retr
+}
+
+func main() {
+	// Prevent log from printing out time information
+	log.SetFlags(0)
+
+	var projectID, bucket, source, name string
+	var public bool
+
+	flag.StringVar(&bucket, "bucket", "", "the bucket to upload content to")
+	flag.StringVar(&projectID, "project-id", "", "the ID of the GCP project to use")
+	flag.StringVar(&source, "source", "", "the path to the source")
+	flag.StringVar(&name, "name", "", "the name of the file to be stored on GCS")
+	flag.BoolVar(&public, "public", true, "whether the item should be available publicly")
+	flag.Parse()
+
+	// If they haven't set the bucket or projectID nor specified
+	// in the environment, then fail if missing.
+	bucket = envOrFatal("GOLANG_SAMPLES_BUCKET", bucket)
+	projectID = envOrFatal("GOLANG_SAMPLES_PROJECT_ID", projectID)
+
+	client, err := defaultGCSClient()
+	if err != nil {
+		log.Fatalf("creating storage service: %v", err)
+	}
+
+	var r io.Reader
+	if source == "" {
+		r = os.Stdin
+		log.Printf("Reading from stdin...")
+	} else {
+		f, err := os.Open(source)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer f.Close()
+
+		r = f
+	}
+	if name == "" {
+		if source != "" {
+			name = filepath.Base(source)
+		} else {
+			name = "test-sample"
+		}
+	}
+
+	_, objAttrs, err := Upload(client, r, projectID, bucket, name, true)
+	if err != nil {
+		log.Fatalf("upload err: %v", err)
+	}
+	log.Printf("URL: %s\nSize: %d\nMD5: %x\n", objectURL(objAttrs), objAttrs.Size, objAttrs.MD5)
+}
+
+func objectURL(objAttrs *storage.ObjectAttrs) string {
+	return fmt.Sprintf("https://storage.googleapis.com/%s/%s", objAttrs.Bucket, objAttrs.Name)
+}
+
+// Upload performs the actual upload to your project's GCS bucket
+func Upload(client *storage.Client, r io.Reader, projectID, bucket, name string, public bool) (*storage.ObjectHandle, *storage.ObjectAttrs, error) {
+	bBucket, _, err := EnsureBucketExists(client, projectID, bucket)
+	if err != nil {
+		return nil, nil, err
+	}
+	ctx := context.Background()
+	obj := bBucket.Object(name)
+	w := obj.NewWriter(ctx)
+	if _, err := io.Copy(w, r); err != nil {
+		return nil, nil, err
+	}
+	if err := w.Close(); err != nil {
+		return nil, nil, err
+	}
+
+	if public {
+		if err := obj.ACL().Set(ctx, storage.AllUsers, storage.RoleReader); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	attrs, err := obj.Attrs(ctx)
+	return obj, attrs, err
+}
+
+func defaultGCSClient() (*storage.Client, error) {
```

Comment:
Yeah am gonna scrap it, it was isolated because it was being invoked by two different tests. Am getting rid of the second test anyways and since I'll be reusing contexts, am gonna scrap it. Thanks.